File: web/src/controller.js
// controller.js — the shed/restore policy. Port of controller.py.
import { SOC_SHED_THRESHOLD, SOC_RESTORE_THRESHOLD, MIN_ACTION_DWELL_HOURS,
        LoadPriority } from "./config.js";

/** Python's max(seq, key=f) returns the FIRST maximum; min the first minimum.
 *  A reduce written with >= or <= returns the LAST, which silently changes
 *  which load is chosen whenever two share a priority. Strict comparison
 *  keeps the first, matching Python. */
function firstBy(items, key, better) {
  let best = null;
  for (const item of items) {
    if (best === null || better(key(item), key(best))) best = item;
  }
  return best;
}
const lowestPriorityFirst = (items) =>
  firstBy(items, (l) => l.priority, (a, b) => a > b);   // max()
const highestPriorityFirst = (items) =>
  firstBy(items, (l) => l.priority, (a, b) => a < b);   // min()

export class AutonomousController {
  constructor({ name = "Autonomous Controller",
                shedThreshold = SOC_SHED_THRESHOLD,
                restoreThreshold = SOC_RESTORE_THRESHOLD,
                minDwellHours = MIN_ACTION_DWELL_HOURS } = {}) {
    Object.assign(this, { name, shedThreshold, restoreThreshold, minDwellHours });
    this.lastChangeH = new Map();
  }

  /** Has this load been left alone long enough to act on again? Per-load. */
  dwellElapsed(load, tHours) {
    if (!this.lastChangeH.has(load.name)) return true;
    return tHours - this.lastChangeH.get(load.name) >= this.minDwellHours;
  }

  /** At most ONE action per tick: a staircase, not a cliff. */
  update(tHours, aggregateSoc, loads, headroomW) {
    // 1. POWER emergency. Dwell is BYPASSED: the bus is already failing, and
    //    an uncontrolled brownout drops loads in whatever order physics picks.
    if (headroomW < 0.0) {
      const candidates = loads.filter(
        (l) => !l.shed && l.priority !== LoadPriority.CRITICAL);
      if (!candidates.length) return null;
      const target = lowestPriorityFirst(candidates);
      target.shed = true;
      this.lastChangeH.set(target.name, tHours);
      return target;
    }

    // 2. ENERGY low. Supply still meets demand, so dwell is enforced.
    if (aggregateSoc < this.shedThreshold) {
      const candidates = loads.filter(
        (l) => !l.shed && l.priority !== LoadPriority.CRITICAL
              && this.dwellElapsed(l, tHours));
      if (!candidates.length) return null;
      const target = lowestPriorityFirst(candidates);
      target.shed = true;
      this.lastChangeH.set(target.name, tHours);
      return target;
    }

    // 3. RECOVERY. BOTH signals must agree: reserves above the restore
    //    threshold AND the load actually fits in the measured headroom.
    if (aggregateSoc > this.restoreThreshold) {
      const candidates = loads.filter(
        (l) => l.shed && this.dwellElapsed(l, tHours)
              && l.demand(tHours) <= headroomW);
      if (!candidates.length) return null;
      const target = highestPriorityFirst(candidates);
      target.shed = false;
      this.lastChangeH.set(target.name, tHours);
      return target;
    }

    return null;   // inside the hysteresis dead band: deliberately idle
  }
}

File: firmware/src/controller.cpp
// controller.cpp — the policy. See controller.hpp for the order it runs in.
#include "controller.hpp"

namespace lunar {

bool Controller::dwell_elapsed(int index, double t_hours) const {
    const double last = last_change_h_[index];
    if (last == NO_CHANGE) return true;
    return t_hours - last >= min_dwell_hours_;
}

void Controller::reset() {
    for (int i = 0; i < MAX_LOADS; ++i) last_change_h_[i] = NO_CHANGE;
}

Action Controller::update(double t_hours, double aggregate_soc,
                          double headroom_w, const LoadView *loads, int count) {
    if (count > MAX_LOADS) count = MAX_LOADS;

    // 1. POWER emergency. Dwell is deliberately not consulted — the bus is
    //    already failing to serve what is connected to it.
    if (headroom_w < 0.0) {
        int target = -1;
        for (int i = 0; i < count; ++i) {
            if (loads[i].shed || loads[i].priority == PRIORITY_CRITICAL) continue;
            // Python's max() keeps the FIRST maximum on ties. Strict > here
            // reproduces that; >= would silently pick the last instead, and
            // the difference only shows up on a tie.
            if (target < 0 || loads[i].priority > loads[target].priority) target = i;
        }
        if (target < 0) return Action{};
        last_change_h_[target] = t_hours;
        return Action{target, true};
    }

    // 2. ENERGY low. Reserves draining, but supply still meets demand.
    if (aggregate_soc < shed_threshold_) {
        int target = -1;
        for (int i = 0; i < count; ++i) {
            if (loads[i].shed || loads[i].priority == PRIORITY_CRITICAL) continue;
            if (!dwell_elapsed(i, t_hours)) continue;
            if (target < 0 || loads[i].priority > loads[target].priority) target = i;
        }
        if (target < 0) return Action{};
        last_change_h_[target] = t_hours;
        return Action{target, true};
    }

    // 3. RECOVERY. A load returns only if it FITS in the measured margin —
    //    never merely because no shortfall is showing, which is precisely what
    //    a successful shed produces.
    if (aggregate_soc > restore_threshold_) {
        int target = -1;
        for (int i = 0; i < count; ++i) {
            if (!loads[i].shed) continue;
            if (!dwell_elapsed(i, t_hours)) continue;
            if (loads[i].demand_w > headroom_w) continue;
            // min() for restore: SMALLEST priority value is the MOST important
            // load, and it comes back first.
            if (target < 0 || loads[i].priority < loads[target].priority) target = i;
        }
        if (target < 0) return Action{};
        last_change_h_[target] = t_hours;
        return Action{target, false};
    }

    return Action{};    // inside the hysteresis dead band: deliberately idle
}

}  // namespace lunar
